rpcs3/Emu/Cell/lv2/sys_uart.cpp

```diff
@@ -1,35 +1,597 @@
+		{
+			u32 rem_size = data_sz;
+
+			while (rem_size)
+			{
+				if (ppu.is_stopped())
+				{
+					return false;
+				}
+
+				std::unique_lock<shared_mutex> lock(vuart_thread.tx_rdy_m);
+				if (vuart_thread.get_tx_bytes() >= PS3AV_TX_BUF_SIZE)
+				{
+					vuart_thread.tx_rdy_c.wait_unlock(5000, lock);
+				}
+				else
+				{
+					lock.unlock();
+				}
+				rem_size -= vuart_thread.enque_tx_data(data + data_sz - rem_size, rem_size);
+			}
+
+			return true;
+		};
+
+		if (!vuart_send_all(data.get(), std::min(ITER_SIZE, static_cast<u32>(size)))) return {};
+
+		if (size > ITER_SIZE)
+		{
+			u32 sent_size;
+
+			for (sent_size = ITER_SIZE; sent_size + ITER_SIZE < size; sent_size += ITER_SIZE)
+			{
+				if (!vuart_send_all(data.get() + sent_size, ITER_SIZE)) return {};
+			}
+
+			if (!vuart_send_all(data.get() + sent_size, static_cast<u32>(size) - sent_size)) return {};
+		}
+
+		ppu.check_state();
+	}
+	else if (mode & 2)
+	{
+		if (!lock.try_lock())
+		{
+			return CELL_EBUSY;
+		}
+
+		if (PS3AV_TX_BUF_SIZE - vuart_thread.get_tx_bytes() < size)
+		{
+			return CELL_EAGAIN;
+		}
+
+		return not_an_error(vuart_thread.enque_tx_data(data.get(), static_cast<u32>(size)));
+	}
+	else
+	{
+		if (!lock.try_lock())
+		{
+			return CELL_EBUSY;
+		}
+
+		const u32 first_iter_sz   = std::min(ITER_SIZE, static_cast<u32>(size));
+		const u32 first_iter_sent = vuart_thread.enque_tx_data(data.get(), first_iter_sz);
+
+		if (first_iter_sent < first_iter_sz || size <= ITER_SIZE)
+		{
+			return not_an_error(first_iter_sz);
+		}
+
+		u32 sent_size;
+
+		for (sent_size = ITER_SIZE; sent_size + ITER_SIZE < size; sent_size += ITER_SIZE)
+		{
+			const u32 in_loop_sent = vuart_thread.enque_tx_data(data.get() + sent_size, ITER_SIZE);
+
+			if (in_loop_sent < ITER_SIZE)
+			{
+				return not_an_error(sent_size + in_loop_sent);
+			}
+		}
+
+		vuart_thread.enque_tx_data(data.get() + sent_size, static_cast<u32>(size) - sent_size);
+	}
+
+	return not_an_error(size);
 }
 
-error_code sys_uart_get_params(vm::ptr<char> buffer)
+error_code sys_uart_get_params(vm::ptr<vuart_params> buffer)
 {
-	sys_uart.todo("sys_uart_get_params(buffer=0x%x)", buffer);
+	sys_uart.trace("sys_uart_get_params(buffer=0x%x)", buffer);
+
+	auto &vuart_thread = g_fxo->get<vuart_av>();
+
+	if (!vuart_thread.init)
+	{
+		return CELL_ESRCH;
+	}
+
+	if (!buffer || !vm::check_addr(buffer.addr(), vm::page_writable, sizeof(vuart_params)))
+	{
+		return CELL_EFAULT;
+	}
+
+	buffer->rx_buf_size = PS3AV_RX_BUF_SIZE;
+	buffer->tx_buf_size = PS3AV_TX_BUF_SIZE;
 
 	return CELL_OK;
 }
+
+void vuart_av_thread::operator()()
+{
+	while (thread_ctrl::state() != thread_state::aborting)
+	{
+		std::unique_lock<shared_mutex> lock(tx_wake_m);
+		if (!tx_buf.get_used_size())
+		{
+			tx_wake_c.wait_unlock(-1, lock);
+		}
+		else
+		{
+			lock.unlock();
+		}
+
+		auto read_size = read_tx_data(temp_tx_buf, sizeof(temp_tx_buf));
+		if (!read_size) continue;
+
+		[&]()
```

Comment:
Consider refactoring this section into its own separate function which would clean up this main loop and avoid the inline lambda that is not reusable anyway.

Reply:
Done